hazelcast/src/main/java/com/hazelcast/topic/MessageListener.java

```diff
@@ -20,23 +20,23 @@
 
 /**
  * Message listener for {@link ITopic}.
+ * <p>
+ * A MessageListener will never be called concurrently (provided that it's not registered twice). So there
+ * is no need to synchronize access to the state it reads or writes. Also there is no need to synchronize
```

Comment:
Will it always be called from the same thread?

Reply:
Not necessarily from the same, but there will be happens-before relation between the previous and next calls.

Reply:
Ok, but if can be called from different threads than the shared state needs to be protected, right?

Reply:
Nope if the environment ensures h-b relation...
For example
```java
int[] value = {10};
Thread t = new Thread(() -> {value[0] = 20;});
t.start();
t.join();
System.out.println(value[0]);
```
`value[0]` is accessed from 2 threads in this piece of code, but doesn't need to be synchronized or volatile. Access from 2 threads doesn't imply it's concurrent.